src/compiler/checker.ts

```diff
@@ -10114,6 +10124,12 @@ namespace ts {
                         // reduced we'll never reduce again, so this occurs at most once.
                         result = getIntersectionType(typeSet, aliasSymbol, aliasTypeArguments);
                     }
+                    else if (extractIrreducible(typeSet, TypeFlags.Undefined)) {
+                        result = getUnionType([getIntersectionType(typeSet), undefinedType], UnionReduction.Literal, aliasSymbol, aliasTypeArguments);

```

Comment:
shouldn't we run both, for large intersections of the form `x | null | undefined & ...`?

Reply:
we do - note the recursive `getIntersectionType` call (which is needed to recalculate includes for the now filtered type set)